### PR TITLE
[WIP] TOML saving and loading hyperparams

### PR DESCRIPTION
**Pull request by Het-Shah**
Currently i have added  only saving feature for the hyperparams from the `get_hyperparams` method. Will add loading.

#249

### CODE REVIEW

**Comment by Sharad24** [comment]
Lets wrap this up soon?

**Comment by Het-Shah** [comment]
I don't think we need to use `getattr` as well. `get_hyperparams` should work just fine  then.

**Comment by Sharad24** [comment]
Ok sounds good

**Comment by Het-Shah** [comment]
Review this? @Sharad24

File: genrl/agents/deep/a2c/a2c.py [hunk 1]
```diff
@@ -193,17 +194,16 @@ def get_hyperparams(self) -> Dict[str, Any]:
             "lr_policy": self.lr_policy,
             "lr_value": self.lr_value,
             "rollout_size": self.rollout_size,
-            "weights": self.ac.state_dict(),
         }
-        return hyperparams
+        return hyperparams, self.ac.state_dict()
 
-    def load_weights(self, weights) -> None:
+    def _load_weights(self, weights) -> None:
```
Comment (Sharad24) [inline genrl/agents/deep/a2c/a2c.py:200]: Any reason to add the `_`?

**Review by Sharad24 — APPROVED** [review]
Looks good. Few things :-

 1. Can you add a short tutorial file on how to load 
        - params?
        - weights?
        - both?
2. Can you add a test for this?

File: genrl/agents/deep/a2c/a2c.py (hunk 1 above)
Comment (Het-Shah) [reply genrl/agents/deep/a2c/a2c.py:200]: it has the same name as the parameter and we are not going to let user get to this :P

**Comment by Het-Shah** [comment]
>Can you add a short tutorial file on how to load
- params?
- weights?
- both?
Can you add a test for this?


Yea sure I'll add these

**Comment by Sharad24** [comment]
CI is failing?

**Comment by Het-Shah** [comment]
Linting is failing wait i'll do that.

**Comment by Het-Shah** [comment]
Wait don't merge this one, i did not change the bandit functions i guess

**Comment by Het-Shah** [comment]
Looks ready now @Sharad24

**Comment by Sharad24** [comment]
Build is failing, tried re-running it -- doesnt work. Can you look into this?

**Comment by Het-Shah** [comment]
=================================== FAILURES ===================================
___________________ TestCBAgent.test_linear_posterior_agent ____________________

self = <tests.test_bandit.test_cb_agents.TestCBAgent object at 0x13d906090>

    def test_linear_posterior_agent(self) -> None:
>       self._test_fn(LinearPosteriorAgent)

tests/test_bandit/test_cb_agents.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_bandit/test_cb_agents.py:35: in _test_fn
    trainer.train(timesteps=10, update_interval=2, update_after=5, batch_size=2)
genrl/trainers/bandit.py:182: in train
    action, kwargs.get("batch_size", 64), train_epochs
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <genrl.agents.bandits.contextual.linpos.LinearPosteriorAgent object at 0x13d659390>
action = tensor(0, dtype=torch.int32), batch_size = 2, train_epochs = 20

    def update_params(
        self, action: int, batch_size: int = 512, train_epochs: Optional[int] = None
    ):
        """Update parameters of the agent.

    
        Updated the posterior over beta though bayesian regression.
    
        Args:
            action (int): Action to update the parameters for.
            batch_size (int, optional): Size of batch to update parameters with.
                Defaults to 512
            train_epochs (Optional[int], optional): Epochs to train neural network for.
                Not applicable in this agent. Defaults to None
        """
        self.update_count += 1
    
        x, y = self.db.get_data_for_action(action, batch_size)
        x = torch.cat([x, torch.ones(x.shape[0], 1)], dim=1)
        inv_cov = torch.mm(x.T, x) + self.lambda_prior * torch.eye(self.context_dim + 1)
>       cov = torch.inverse(inv_cov)
E       RuntimeError: inverse_cpu: U(6,6) is zero, singular U.

genrl/agents/bandits/contextual/linpos.py:148: RuntimeError
----------------------------- Captured stdout call -----------------------------

Started at 07-09-20 14:41:35
Training LinearPosteriorAgent on CovertypeDataBandit for 10 timesteps


This is the error trace i don't really know what this is?

**Comment by Sharad24** [comment]
Can you sync up this branch with the master? I think this was an issue a few days back, but was resolved.

**Comment by Het-Shah** [comment]
Yea cool

**Comment by Het-Shah** [comment]
Still getting it?

**Comment by Sharad24** [comment]
Fetch and Merge again, the test was removed in #329

**Pull request merged**